app/moderation-algo/main.py

```diff
@@ -0,0 +1,261 @@
+from waitress import serve
+from TopicExtraction.analyzeReadingMaterials import ExtractIssueFromReadingMaterials, ExtractTopicsDescriptions
+from TopicExtraction.extractTopicsInteractive import Lock, LockMany, Unlock, UnlockMany, Rename, Add, Remove, RemoveMany, Regenerate
+from AlgoStages.substantiveness import IsSubstantive
+from AlgoStages.offensiveness import IsOffensive
+from AlgoStages.relevance import IsRelevant
+from AlgoStages.classifyQuestion import DoesQuestionFitCategory
+from Translation.translation import TranslateText
+from PerspectiveAPI import InitPerspectiveAPI
+from Utilities.logEvents import LogEventConsole
+import redis
+import json
+import os
+
+app = Flask(__name__)
+
+def ProcessQuestion(issue: str, topics: dict, question: str) -> dict:
+    """Process a question through the algorithm and return the results from each stage as a dict.
+    @Param: issue = Broad issue that the town hall is about (Ex: 'modernizing congress')
+    @Param: question = User question or comment to process
+    @Return: response = {'substantive': bool, 'offensive': bool, 'relevant': bool, 'topics': {'topic0': bool, 'topic1': bool, ...}}"""
+    # Make sure Google API is initialized
+    model = 'gemini-pro'
+
+    # First translate the question into English and Spanish. The algorithm is built to process on English questions
+    question_en, question_es, original_lang = TranslateText(question)
+
+    # Process the question through the algo and return the following fields
+    response = {
+        'question_en': question_en,
+        'question_es': question_es,
+        'original_lang': original_lang,
+        'substantive': False,
+        'offensive': False,
+        'relevant': False,
+        'topics': {},
+    }
+
+    # Check for substantiveness
+    response['substantive'] = IsSubstantive(model, question_en)
+
+    # Check for offensiveness
+    response['offensive'] = IsOffensive(question_en)
+
+    # Don't bother with further processing if the question is unsubstantive or offensive
+    if(response['offensive'] or not response['substantive']):
+        return response
+
+    # Check for relevancy if an issue was provided
+    if(issue):
+        response['relevant'] = IsRelevant(model, question_en, issue)
+
+    # Classify into topics if any were provided
+    if(topics):
+        for topic, description in topics.items():
+            response['topics'][topic] = DoesQuestionFitCategory(model, question_en, topic, description)
+
+    return response
+
+def ConnectToRedis() -> redis.Redis:
+    "Establish a connection to Redis for temporary persistent memory. Returns the connection object."
+    REDIS_HOST = os.environ.get('REDIS_HOST')
+    REDIS_PORT = os.environ.get('REDIS_PORT')
+    REDIS_USERNAME = os.environ.get('REDIS_USERNAME')
+    REDIS_PASSWORD = os.environ.get('REDIS_PASSWORD')
+    return redis.StrictRedis(host=REDIS_HOST, port=REDIS_PORT, username=REDIS_USERNAME, password=REDIS_PASSWORD,
```

Comment:
This is good, but I believe when deployed it only needs the host and port since the redis cluster is only accessible from within the cluster. Not sure how this behaves if the username and password are undefined. Might be fine, but we may want a check here for the `NODE_ENV`, if it is `production` we can just return it with just the Host and Port provided, otherwise it will stay the same.